### PR TITLE
feat: support DeepSeek V4 context parallel training

### PR DESCRIPTION
**Pull request by HuiyingLi**
## Summary
- add Miles-style context parallel support for DeepSeek V4 non-packed training
- all-gather vanilla KV and compressed KV across DSV4 CP ranks with autograd-aware collectives
- build local-query/global-key CP masks and use global query positions for sparse/window/compressed topk
- wire DSV4 CP through the LLM fine-tuning recipe and model capability checks

## Details
This stacks on #1914's manual all-gather CP path. DSV4 attention does not use Transformer Engine attention modules, so the CP process group is passed into the custom DSV4 attention path instead. Each CP rank owns a contiguous local query shard, while vanilla KV and compressed KV are gathered across the CP group before TileLang sparse attention consumes them.

Packed sequence support will follow in a separate change.

## Validation
- `pytest -q tests/unit_tests/models/deepseek_v4/test_dsv4_cp_distributed.py`
- real 8-GPU `train_ft.py` recipe runs through `torchrun -m nemo_automodel.cli.app`, first 4 real DSV4 layers, EP8, seq_len=4096, global batch size 8, 100 steps:
  - CP1: https://wandb.ai/Nemo-automodel/dsv4-cp-parity/runs/i27xbrhi
    - 100 logged loss points, step 0/99: `20.6404476166` / `13.9647502899`
    - max logged `mem`: `44.0738544464` GiB
  - CP2: https://wandb.ai/Nemo-automodel/dsv4-cp-parity/runs/x6tu0l32
    - 100 logged loss points, step 0/99: `20.6329612732` / `13.9671049118`
    - max logged `mem`: `45.8731751442` GiB
  - CP1 used grad accumulation 1; CP2 used grad accumulation 2 to keep the same effective global batch with DP4/CP2.

### CODE REVIEW

File: nemo_automodel/recipes/llm/train_ft.py [hunk 1]
```diff
@@ -1373,6 +1396,16 @@ def _forward_backward_step(
             )
             for k, v in batch.items()
         }
+        model_for_cp = self.model_parts[0] if hasattr(self, "model_parts") else self.cfg.model
+        cp_size = getattr(getattr(self, "dist_setup", None), "cp_size", self.cfg.get("distributed.cp_size", 1))
+        dsv4_manual_cp = cp_size > 1 and _is_deepseek_v4_model_or_config(model_for_cp)
+        if dsv4_manual_cp:
```
Comment (akoumpa) [inline nemo_automodel/recipes/llm/train_ft.py:1402]: Hi @HuiyingLi, would it be possible to not have model-specific logic here?
